src/workspace_storage.rs

```diff
@@ -0,0 +1,545 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
+use pyo3::{
+    import_exception,
+    prelude::{pyclass, pyfunction, pymethods, IntoPy, PyAny, PyObject, PyResult, Python},
+    types::PyBytes,
+    PyErr,
+};
+use std::{
+    collections::{HashMap, HashSet},
+    path::PathBuf,
+};
+
+use crate::{
+    data::{LocalFileManifest, LocalFolderManifest, LocalUserManifest, LocalWorkspaceManifest},
+    ids::{BlockID, ChunkID, EntryID},
+    local_device::LocalDevice,
+    regex::Regex,
+    time::DateTime,
+};
+
+use libparsec::core_fs::FSError;
+
+import_exception!(parsec.core.fs.exceptions, FSLocalMissError);
+import_exception!(parsec.core.fs.exceptions, FSInvalidFileDescriptor);
+import_exception!(parsec.core.fs.exceptions, FSLocalStorageOperationalError);
+import_exception!(parsec.core.fs.exceptions, FSLocalStorageClosedError);
+import_exception!(parsec.core.fs.exceptions, FSInternalError);
+
+/// WorkspaceStorage's binding is implemented with allow_threads because its
+/// methods are called in trio.to_thread to connect the sync and async world
+#[pyclass(module = "parsec._parsec")]
+#[derive(Clone)]
+pub(crate) struct WorkspaceStorage(pub libparsec::core_fs::WorkspaceStorage, pub Option<Regex>);
+
+#[pymethods]
+impl WorkspaceStorage {
+    #[new]
+    #[args(cache_size = "512 * 1024 * 1024")]
+    #[allow(unused_variables)]
+    fn new(
+        data_base_dir: &PyAny,
+        device: LocalDevice,
+        workspace_id: EntryID,
+        prevent_sync_pattern: Regex,
+        cache_size: u64,
+        data_vacuum_threshold: Option<u64>,
+    ) -> PyResult<Self> {
+        let data_base_dir =
+            PathBuf::from(data_base_dir.call_method0("__str__")?.extract::<String>()?);
```

Comment:
You can extract &str, if you don't need ownership

Reply:
done